Bluetooth: Add HCI User Channel driver for native POSIX port

jhedberg opened the pull request:
Introduce a custom HCI driver for the native POSIX port, which opens a
HCI User Channel socket to the Linux kernel to gain access to a local
Bluetooth controller.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

aescolar-ot commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+	 *
+	 *  - Iterate all adapters and pick the first non-powered one
+	 *  - Use an environment variable to specify adapter
+	 *  - Use a command line option (e.g. --bt-dev hci0)
```
I'd think the command line option would be the best option. You can add it easily already, and you can have a default one. I could add a patch for you if you'd want.

holtmann replied:
Maybe the simplest is to use a mandatory --bt-dev command line option. Since there is really no easy choice for an adapter, force the user to specify one.

carlescufi replied:
I agree, it's better to make it mandatory in order to avoid confusion

aescolar-ot commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+#define H4_SCO           0x03
+#define H4_EVT           0x04
+
+static BT_STACK_NOINIT(rx_thread_stack, CONFIG_BT_RX_STACK_SIZE);
```
You only need CONFIG_ARCH_POSIX_RECOMMENDED_STACK_SIZE.
The underlying (host) pthread has another stack. This zephyr stack does not hold the real thread stack.

jhedberg replied:
Thanks. I'll fix this in the next revision.

aescolar-ot commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+		if (len < 0) {
+			BT_ERR("Reading socket failed, errno %d", errno);
+			k_yield();
+			continue;
```
is this something you can actually recover from?

jhedberg replied:
At least in some cases yes, the syscall might return due to an interrupt and set errno = EINTR.

aescolar-ot commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+
+		BT_DBG("calling read()");
+
+		len = read(uc_fd, frame, sizeof(frame));
```
I'm not used to BT sockets, so probably this is a silly question: Can you assume you will always get full HCI packets? (meaning, not segmented in the socket itself)

jhedberg replied:
Not all Bluetooth sockets are like this, but User Channel sockets are. I.e. they're packet based and always take and give full frames.

aescolar reviewed (commented):
Nice. Just some minor comments/questions from my side.

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+		ssize_t len;
+
+		if (!uc_ready()) {
+			k_sleep(K_MSEC(20));
```
@aescolar-ot any comments on the need to do the above? Ideally I'd just want to do blocking read() calls on the socket from this thread and have the blocking call act as k_yield() if there's currently no data. Having to do constant looping with a small timeout is inefficient and not very pretty. I've noticed that both the console and network drivers take the same approach, so I assumed this is the only option for now.

aescolar-ot replied:
That call is to the underlying host OS, so if it were to block, it would leave blocked all the Zephyr threads. To do it more efficiently I would need to probably provide you a replacement for the host read() in the arch which would do the magic, but it would complicate it quit a bit. So unfortunately, no, there is nothing better by now.

jhedberg replied:
I've added this as an item to #6044. If it's at all technically doable, I think it'd be a very valuable addition. If you have to do this separately for each of the different function calls, I'd start by doing it for select or poll, since that will then cover both read and write, as well as handling multiple file descriptors.

holtmann commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -0,0 +1,214 @@
+
+#define SOL_HCI          0
+
+#define H4_NONE          0x00
```
What is H4_NONE and why it is needed?

jhedberg commented:
I think I've addressed all comments now, except for the command line option handling. @aescolar-ot what's the right way to include the header files in `boards/posix/native_posix/*.h`? I suspect that path would need to be added as an include lookup for `drivers/bluetooth/hci/userchan.c`, but I'm not sure how to do that (I found samples on how to add a new include lookup location for an entire directory, but not for a single c-file)

aescolar-ot commented:
>  I suspect that path would need to be added as an include lookup for ...

I don't think you need to. If it is just for the command line parsing, you can add an ifdef in the cmdline.c like for the entropy driver (it is not too beautiful but it's easy enough)

carlescufi commented on boards/posix/native_posix/native_posix_defconfig:
```diff
@@ -6,3 +6,4 @@ CONFIG_SERIAL=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=1000000
 CONFIG_COVERAGE=y
 CONFIG_FAKE_ENTROPY_NATIVE_POSIX=y
+CONFIG_BT_USERCHAN=y
```
If we make this part of the default board and require a command-line option this might not work for those not interested in Bluetooth using the native port?

jhedberg replied:
I think it should be fine, since the option depends on CONFIG_BT, i.e. the above enabling will never happen if the pre-requisite is not fulfilled. E.g. I just tried building hello_world for native_posix, and not a single CONFIG_BT* option ended up in the final configuration.

jhedberg replied:
Also, I'd make the check for the presence of the command line option to happen only when bt_enable() (and consequently drv->open) is called.

Commit:
Bluetooth: Add HCI User Channel driver for native POSIX port

Introduce a custom HCI driver for the native POSIX port, which opens a
HCI User Channel socket to the Linux kernel to gain access to a local
Bluetooth controller.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
Updated now to support a --bt-dev=hci0 command line switch (hence I removed the [RFC] tag). There's a new `extern int bt_dev_index` variable exposed through the native_posix board.h.

aescolar-ot commented on boards/posix/native_posix/cmdline.c:
```diff
@@ -36,6 +38,19 @@ static void cmd_seed_found(char *argv, int offset)
+static void cmd_bt_dev_found(char *argv, int offset)
+{
+	if (strncmp(&argv[offset], "hci", 3) || strlen(&argv[offset]) < 4) {
+		return;
```
Would it make sense to be a bit kinder to users and print a warning? (if things do not break after, or an error if they do). Something like:
posix_print_warning("Warning: the bt-dev command line option only accepts hci devices, for ex. --bt-dev:hci0 (you provided %s)\n", argv);

jhedberg replied:
That's a good suggestion. I'll add it.

aescolar-ot commented on boards/posix/native_posix/cmdline.c:
```diff
@@ -66,6 +81,13 @@ void native_handle_cmd_line(int argc, char *argv[])
 #endif
 
+#if defined(CONFIG_BT_USERCHAN)
+		{ false, false, false,
```
If the parameter is mandatory, maybe better set mandatory to true (the 2nd parameter in this line), so the help message reflects it.

jhedberg replied:
The way I was thinking, is that the errors would only come once the application decides it wants to use Bluetooth, i.e. calls the bt_enable() API (which will return an error, which all the Bluetooth-using samples will print to console). I'm not convinced that this is the best approach though, i.e. the assumption that there may apps that enable Bluetooth in Kconfig but never call bt_enable().

aescolar-ot replied:
That mandatory switch only affects the syntax displayed in the help message. So you can still throw the error at any point later (or never).
If you want the error displayed ASAP, you could print it after the for loop in `native_handle_cmd_line()` if the option was not found.
If you want to postpone it to the driver init you can also do that.
Anyhow, enabling bluetooth but never using it would seem more like a misconfiguration.

aescolar-ot commented:
Would you have a bit of time to add a couple of paragraphs in https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/native_posix/doc/board.rst#peripherals ?

aescolar-ot commented:
Another minor thing, as the BT shell app sets `TEST` , the time will not be slowed down to real time, which will give a quite weird behaviour in this use case. So you may want to apply the patch below
```diff
diff --git a/boards/posix/native_posix/Kconfig b/boards/posix/native_posix/Kconfig
index 9ee31d5..808b995 100644
--- a/boards/posix/native_posix/Kconfig
+++ b/boards/posix/native_posix/Kconfig
@@ -6,8 +6,8 @@ comment "Native POSIX options"
 config NATIVE_POSIX_SLOWDOWN_TO_REAL_TIME
        bool "Slow down execution to real time"
        depends on BOARD_NATIVE_POSIX
-       default y if !TEST
-       default n if TEST
+       default y if (!TEST) || BT_USERCHAN
+       default n if TEST && (!BT_USERCHAN)
        help
          When selected the execution of the process will be slowed down to real time.
          (if there is a lot of load it may be slower than real time)
```
Anyhow, I just added another minor PR (#7943) so it can be controlled from command line, without the need to set options.

jhedberg commented:
> Would you have a bit of time to add a couple of paragraphs in https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/native_posix/doc/board.rst#peripherals ?

Sure, I'll add a section there.

jhedberg commented:
> Another minor thing, as the BT shell app sets TEST , the time will not be slowed down to real time, which will give a quite weird behaviour in this use case. So you may want to apply the patch below

I have a feeling it's a mistake that `tests/bluetooth/shell` enables `CONFIG_TEST`. Grepping the tree for its usage gives the impression that it's intended for self-contained automatest tests, which the Bluetooth shell is not (it's a manual tool requiring a human to operate it). So I'd instead consider removing `CONFIG_TEST=y` from the Bluetooth shell. @nashif does this sound reasonable or I have the wrong understanding of what `CONFIG_TEST` is for?

jhedberg commented:
@aescolar-ot I've now updated the PR with all your suggestions, except the `CONFIG_TEST` issue (see my other comment about it)

dbkinder commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -411,6 +411,16 @@ The following peripherals are currently provided with this board:
   needs elevated permissions.
 
+**Bluetooth controller**:
+  It's possible to use the host's Bluetooth adapter as a Bluetooth
```
Some formatting and wording tweaks:

    It's possible to use the host's Bluetooth adapter as a Bluetooth
    controller for Zephyr. To do this, the HCI device needs to be
    passed as a command line option to ``zephyr.exe``. For example,
    to use ``hci0``, use ``sudo zephyr.ext --bt-dev=hci0``.  Using
    the device requires root privileges (or the CAP_NET_ADMIN POSIX
    capability, to be exact)  so ``zephyr.exe`` needs to be through
    ``sudo``. The chosen HCI device must be powered down and
    support Bluetooth Low Energy (Bluetooth specification version
    4.0 or greater).

jhedberg replied:
fixed

jhedberg commented:
>> Another minor thing, as the BT shell app sets TEST , the time will not be slowed down to real time, which will give a quite weird behaviour in this use case. So you may want to apply the patch below

> I have a feeling it's a mistake that tests/bluetooth/shell enables CONFIG_TEST. Grepping the tree for its usage gives the impression that it's intended for self-contained automatest tests, which the Bluetooth shell is not (it's a manual tool requiring a human to operate it). So I'd instead consider removing CONFIG_TEST=y from the Bluetooth shell. @nashif does this sound reasonable or I have the wrong understanding of what CONFIG_TEST is for?

@nashif any comment on the above? Since CONFIG_TEST influences the "real time" behavior of the native port, it seems like it's the wrong thing to enable for any Bluetooth app (but especially the shell one), since all of them have various timers, etc that will break interoperability when talking to remote devices (which expect correct timings).

dbkinder reviewed (approved):
+1 for docs, thanks!
